src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -3205,15 +3205,18 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           //         the corresponding synthetics to the package class, only to the package object class.
           // scala/bug#6734 Locality test below is meaningless if we're not even in the correct tree.
           //         For modules that are synthetic case companions, check that case class is defined here.
+          // scala/bug#10783 ditto for synthetic companions of derived value classes.
           def shouldAdd(sym: Symbol): Boolean = {
+            def classDefinedHere(s: Symbol): Boolean = stats exists {
+              case t: ClassDef => t.symbol eq s
+              case _           => false
+            }
             def shouldAddAsModule: Boolean =
               sym.moduleClass.attachments.get[ClassForCaseCompanionAttachment] match {
                 case Some(att) =>
-                  val cdef = att.caseClass
-                  stats.exists {
-                    case t @ ClassDef(_, _, _, _) => t.symbol == cdef.symbol   // cdef ne t
-                    case _ => false
-                  }
+                  classDefinedHere(att.caseClass.symbol)
+                case _ if sym.companionClass.isDerivedValueClass =>
+                  classDefinedHere(sym.companionClass)
```

Comment:
It seems wrong to need a special case for value classes here, rather than being able to generically detect that `ensureCompanionObject` was called and no explicit companion was in the vicinity.

e.g. default arguments for constructors give rise to synthetic companions to host the default getters. And compiler plugins might use `ensureCompanionObject` for custom, but similar, use cases.

Reply:
The default getters case was mentioned by contributor. Isn't the general solution to use attachments for all trees with associated synthetics? Currently we have synthetic syms and ask if they belong in the current stats, instead of looking at the stats for things with synthetics. In my second contribution, I punned on ad-hoc as add-hack, that was a good one. https://github.com/scala/scala/commit/ed723738e643035609efda49e6be689e4b0b0480

Reply:
(Obviously, I wrote my reply before I saw hrhino's.)